plugins/chat/app/models/chat/message.rb

```diff
@@ -91,11 +95,56 @@ class Message < ActiveRecord::Base
 
     before_save { ensure_last_editor_id }
 
+    normalizes :blocks,
+               with: ->(blocks) do
+                 return if !blocks
+
+                 # automatically assigns unique IDs
+                 blocks.each do |block|
+                   block["block_id"] ||= SecureRandom.uuid
+                   block["elements"].each do |element|
+                     element["action_id"] ||= SecureRandom.uuid if element["type"] == "button"
+                   end
+                 end
+               end
+
     validates :cooked, length: { maximum: 20_000 }
-    validate :validate_message
+
+    validate :validate_blocks
+    def validate_blocks
+      return if !blocks
+
+      schemer = JSONSchemer.schema(Chat::Schemas::MessageBlocks)
+      if !schemer.valid?(blocks)
+        errors.add(:blocks, schemer.validate(blocks).map { _1.fetch("error") })
+        return
+      end
+
+      block_ids = Set.new
+      action_ids = Set.new
+      blocks.each do |block|
+        block_id = block["block_id"]
+        if block_ids.include?(block_id)
+          errors.add(:blocks, "have duplicated block_id: #{block_id}")
+          next
+        end
+        block_ids.add(block_id)
+
+        block["elements"].each do |element|
+          action_id = element["action_id"]
+          next unless action_id
+          if action_ids.include?(action_id)
+            errors.add(:elements, "have duplicated action_id: #{action_id}")
+            next
+          end
+          action_ids.add(action_id)
+        end
+      end
+    end
 
     def self.polymorphic_class_mapping = { "ChatMessage" => Chat::Message }
 
+    validate :validate_message
```

Comment:
Could we keep DSL keywords grouped together instead of spreading them? I know we don’t have clear rules on that, but I think it’s better to easily see all the same related things grouped in one place. Wdyt?